### ISSUE TITLE
Avoid "unsupported by backend contract: module initializers" when an attribute is written to but never read

### ISSUE THREAD

**Issue by mgehre-amd** [llvm/torch-mlir#2187]
The code
```
import torch
import torch_mlir

class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.p = torch.zeros([3])
    
    
    def forward(self):
        torch.ops.aten.uniform_(self.p, 0.0, 1.0)
        return True
        

model = Model()
inps = ()
out = model(*inps)
print(out)
print(torch_mlir.compile(model, inps, output_type='raw'))
torch_mlir.compile(model, inps, output_type='torch')
```
fails with
```
python exception: Failure while executing pass pipeline:
error: unknown: unsupported by backend contract: module initializers
note: unknown: see current operation: "torch.initialize.global_slots"(%2) {slotSymNames = [@p]} : (!torch.tensor<[3],f32>) -> ()
note: unknown: this is likely due to InlineGlobalSlots being unable to inline a global slot
```
The IR is
```
// -----// IR Dump After LowerToBackendContract Failed (torch-lower-to-backend-contract) //----- //
module attributes {torch.debug_module_name = "Model"} {
  torch.global_slot.module_initializer {
    %none = torch.constant.none
    %true = torch.constant.bool true
    %0 = torch.tensor.literal(dense<[0.857172251, 0.710896373, 0.145485401]> : tensor<3xf32>) : !torch.tensor<[3],f32>
    torch.initialize.global_slots [
      @p(%0 : !torch.tensor<[3],f32>)
    ]
  }
  torch.global_slot "private" @p : !torch.tensor
  func.func @forward() -> !torch.bool {
    %int3 = torch.constant.int 3
    %int4 = torch.constant.int 4
    %none = torch.constant.none
    %true = torch.constant.bool true
    %float0.000000e00 = torch.constant.float 0.000000e+00
    %float1.000000e00 = torch.constant.float 1.000000e+00
    %0 = torch.global_slot.get @p : !torch.tensor
    %1 = torch.copy.to_vtensor %0 : !torch.vtensor
    %2 = torch.prim.dtype %1 : !torch.vtensor -> !torch.int
    %3 = torch.prim.ListConstruct %int3, %int4 : (!torch.int, !torch.int) -> !torch.list<int>
    %4 = torch.aten.empty.memory_format %3, %2, %none, %none, %none, %none : !torch.list<int>, !torch.int, !torch.none, !torch.none, !torch.none, !torch.none -> !torch.vtensor<[3,4],unk>
    %5 = torch.aten.uniform %4, %float0.000000e00, %float1.000000e00, %none : !torch.vtensor<[3,4],unk>, !torch.float, !torch.float, !torch.none -> !torch.vtensor<[3,4],unk>
    %6 = torch.tensor_static_info_cast %5 : !torch.vtensor<[3,4],unk> to !torch.vtensor
    torch.overwrite.tensor.contents %6 overwrites %0 : !torch.vtensor, !torch.tensor
    return %true : !torch.bool
  }
}
```

I understand that it's generally complicated to allow models to modify the attributes, but I'm seeing some models that modify attributes which are then unused.
Is there a way to see that `self.p` single use is in `torch.overwrite.tensor.contents` and thus remove the `torch.overwrite.tensor.contents`. This should allow to eventually also drop the `torch.global_slot.get`.

**Comment by vivekkhandelwal1** [comment]
> I understand that it's generally complicated to allow models to modify the attributes, but I'm seeing some models that modify attributes which are then unused.

In general, if you are mutating any input value in place then it results in this error. I'm not sure if such kind of code is lowered in torch-mlir successfully. Do you know of any such example?

**Comment by mgehre-amd** [comment]
Yes, we are seeing this in some transformer models that sanitize their configuration on first invocation, and consequently adjust their model representation. You can reproduce it with the reduced model I attached in this issue.

**Comment by vivekkhandelwal1** [comment]
> Yes, we are seeing this in some transformer models that sanitize their configuration on first invocation, and consequently adjust their model representation. You can reproduce it with the reduced model I attached in this issue.

Yeah, I'm able to reproduce the issue. I am asking for an example where we're mutating a tensor and it's getting lowered successfully via torch-mlir. Do you know of any such example? Because I haven't seen this kind of IR getting lowered successfully via torch-mlir.

**Comment by mgehre-amd** [comment]
> > Yes, we are seeing this in some transformer models that sanitize their configuration on first invocation, and consequently adjust their model representation. You can reproduce it with the reduced model I attached in this issue.
> 
> Yeah, I'm able to reproduce the issue. I am asking for an example where we're mutating a tensor and it's getting lowered successfully via torch-mlir. Do you know of any such example? Because I haven't seen this kind of IR getting lowered successfully via torch-mlir.

There is an example of a tensor being mutated in-place here: https://github.com/llvm/torch-mlir/issues/2165

For the issue here, I expected that a global dead symbol analysis would remove `self.p` and all ops accessing it as `dead`, and then there would be no tensor modifications left in the module.

**Comment by vivekkhandelwal1** [comment]
> 
> There is an example of a tensor being mutated in-place here: #2165

That code is getting lowered because the first operand of the op `torch.ops.aten.add_(convolution_7, _param_constant82)` is `convolution_7` which is an input to the forward function. If you just interchange the order of both the args this will also fail. And you will get the following IR dump:
```
// -----// IR Dump After LowerToBackendContract Failed (torch-lower-to-backend-contract) //----- //
module attributes {torch.debug_module_name = "Model"} {
  torch.global_slot.module_initializer {
    %none = torch.constant.none
    %true = torch.constant.bool true
    %0 = torch.tensor.literal(dense<2.000000e+00> : tensor<1x384x12xf32>) : !torch.tensor<[1,384,12],f32>
    torch.initialize.global_slots [
      @_param_constant82(%0 : !torch.tensor<[1,384,12],f32>)
    ]
  }
  torch.global_slot "private" @_param_constant82 : !torch.tensor
  func.func @forward(%arg0: !torch.vtensor<[1,384,12],f32>) -> !torch.vtensor {
    %int1 = torch.constant.int 1
    %0 = torch.global_slot.get @_param_constant82 : !torch.tensor
    %1 = torch.copy.to_vtensor %0 : !torch.vtensor
    %2 = torch.aten.add.Tensor %1, %arg0, %int1 : !torch.vtensor, !torch.vtensor<[1,384,12],f32>, !torch.int -> !torch.vtensor
    torch.overwrite.tensor.contents %2 overwrites %0 : !torch.vtensor, !torch.tensor
    %3 = torch.copy.to_vtensor %0 : !torch.vtensor
    return %3 : !torch.vtensor
  }
}
```
And here you can see that it's trying to mutate the model attr and hence it failed. 

>
> For the issue here, I expected that a global dead symbol analysis would remove self.p and all ops accessing it as dead, and ?> then there would be no tensor modifications left in the module.

Since you are modifying a model attr in place even if that attr is not used within the forward function which torch-mlir is compiling, it could be a case that the same attr value could be used at some other place in a separate function. So ideally that's not a dead code if we talk in the scope of the entire class. Because the attr can still have uses outside the forward method, which would get affected if we eliminate this as a dead code.

@ramiro050 can comment more on this and correct me if I'm wrong.

**Comment by ramiro050** [comment]
I agree with @vivekkhandelwal1. The mutation of the slot is a side-effect of `forward` that lives on after `forward` is done executing, so it would not be correct to remove the mutation even if the tensor is not returned by `forward`.

**Comment by mgehre-amd** [comment]
I see. So I guess this case will never be supported by torch-mlir, correct?

**Comment by ramiro050** [comment]
I think torchdynamo is [expected to always produce](https://github.com/llvm/torch-mlir/blob/c3cd7471b4831d101780fdd9527feb7cce907bc3/python/torch_mlir/_dynamo_fx_importer.py#L436-L438) stateless FX graphs, so this sort of problem should go away once torchdynamo becomes a viable path. Since our long term roadmap is to fully align with torchdynamo, I don't expect support for keeping track of state to take place. But if it is a big blocker for you, we can definitely explore options to deal with this.
